add tutorial for masks (#260)
histolab/histolab: docs/tissue_masks.rst

=======================

When extracting the tile dataset from the WSI collection, we may want to consider only a portion of the tissue, rather than the whole slide.
For example, a single WSI can include multiple individual slices of tissue, or we may have pathologist annotations with the regions of interest (ROIs) we need to consider.

In this tutorial, we will see how to define different tissue masks in order to refine the tile extraction procedure.

First, we need to load some modules and an example `Slide <slide.html#src.histolab.slide.Slide>`_; here we will consider the `Kidney <data.html#src.histolab.data.ihc_kidney>`_ WSI available in the `data <data.html#data>`_ module.

.. code-block:: ipython3

   from histolab.slide import Slide
   from histolab.data import ihc_kidney
   import os

   BASE_PATH = os.getcwd()

   PROCESS_PATH = os.path.join(BASE_PATH, 'kidney', 'processed')

   ihc_kidney_svs, ihc_kidney_path = ihc_kidney_svs()
   ihc_kidney_slide = Slide(ihc_kidney_path, processed_path=PROCESS_PATH)

   ihc_kidney_slide.thumbnail

.. figure:: https://user-images.githubusercontent.com/31658006/116117338-672d0c00-a6bc-11eb-87a3-1b752aff46d0.png
   :alt: ihc_kidney

> On our `Slide <slide.html#src.histolab.slide.Slide>`_ object we can now retrieve a binary mask considering specific regions of the tissue. Notice that available masks are defined in the `masks <masks.html#src.histolab.masks>`_ module.

Reviewer: ```suggestion
From our `Slide <slide.html#src.histolab.slide.Slide>`_ object we can now retrieve a binary mask considering specific regions of the tissue. Notice that available masks are defined in the `masks <masks.html#src.histolab.masks>`_ module.
```

In particular, the `TissueMask <masks.html#src.histolab.masks.TissueMask>`_ accounts for all the tissue detected on the slide, while the `BiggestTissueBoxMask <masks.html#src.histolab.masks.BiggestTissueBoxMask>`_ keeps only the largest connected component of the tissue.

As a diagnostic check to visualize the boundaries of the selected mask, we can call the ``locate_mask`` method on the Slide:

.. code-block:: ipython3

   from histolab.masks import BiggestTissueBoxMask
   largest_area_mask = BiggestTissueBoxMask()
   ihc_kidney_slide.locate_mask(largest_area_mask)

.. figure:: https://user-images.githubusercontent.com/31658006/116119576-e02d6300-a6be-11eb-85b2-01df96c9c3eb.png
   :alt: largest bbox

.. code-block:: ipython3

   from histolab.masks import TissueMask
   all_tissue_mask = TissueMask()
   ihc_kidney_slide.locate_mask(all_tissue_mask)

.. figure:: https://user-images.githubusercontent.com/31658006/116119755-0f43d480-a6bf-11eb-86eb-3f5b933ede1c.png
   :alt: all tissue

> It is also possible to define a custom binary mask via the `BinaryMask <masks.html#src.histolab.masks.BinaryMask>`_ object.

Reviewer: ```suggestion
It is also possible to define a custom binary mask by subclassing the `BinaryMask <masks.html#src.histolab.masks.BinaryMask>`_ object.
```

Reviewer: this was not resolved

For example, we can limit a rectangular region with upper-left coordinates (400, 280) and bottom-right coordinates (300, 320):

.. code-block:: ipython3

   from histolab.masks import BinaryMask
   from histolab.util import polygon_to_mask_array
   from histolab.types import CP

   class MyCustomMask(BinaryMask):
        def _mask(self, slide):
            thumb = slide.thumbnail
            my_mask = polygon_to_mask_array(thumb.size, CP(400, 280, 300, 320))
            return my_mask

   custom_mask = MyCustomMask()

   ihc_kidney_slide.locate_mask(custom_mask)

.. figure:: https://user-images.githubusercontent.com/31658006/116122414-0acceb00-a6c2-11eb-9af7-b948592ab9ec.png
   :alt: all tissue

> We can finally pass our mask to the ``extract`` method of our `Tiler <tiler.html#src.histolab.tiler>`_ object, and visualize the location of the extracted tiles:

Reviewer: please add a link to the `extract` method

.. code-block:: ipython3

    from histolab.tiler import RandomTiler

>     rtiler = RandomTiler(
>                     prefix="",
>                     tile_size=(128,128),
>                     level=0,
>                     n_tiles=50,
>                     check_tissue=True,
>                     tissue_percent=90,
>                     seed=0,
>                 )

Reviewer: ```suggestion
    rtiler = RandomTiler(
        tile_size=(128, 128),
        n_tiles=50,
        level=0,
        tissue_percent=90,
        seed=0,
    )
```

1. fixed indentation
2. reordered params according to "normal" order
3. removed `check_tissue` and `prefix` that were defaults

>     rtiler.extract(ihc_kidney_slide, all_tissue_mask, log_level="WARNING")

Reviewer: ```suggestion
    rtiler.extract(ihc_kidney_slide, all_tissue_mask)
```

removed since this not a tutorial about that

    rtiler.locate_tiles(
        slide=ihc_kidney_slide,

>         extraction_mask = all_tissue_mask,

Reviewer: ```suggestion
        extraction_mask=all_tissue_mask,
```
